회원가입 구현

kihyuk-jeong opened the pull request:
- 회원가입 / 테스트코드 
- 데이터 확인을 위한 findAll 기능 구현(추후 삭제)

kihyuk-jeong committed:
[#2] 회원가입 구현

kihyuk-jeong committed:
[#2] 패키지 구조 변경

Hosick commented on src/main/java/com/flab/shoeauction/common/exception/CustomizeResponseEntityExceptionHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+@RestControllerAdvice
+public class CustomizeResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {
+
```
유효성 검사에서 발생한  MethodArgumentNotValidException은 따로 핸들링 해주시지 않는건가요?

Hosick commented on src/main/java/com/flab/shoeauction/user/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @GetMapping("/duplicated/nickname")
+    public boolean nickNameDuplicated(String nickname) {
```
'nickname' 자체가 하나의 명사인데 카멜케이스로 적는게 맞을까요??

kihyuk-jeong replied:
nicknameDuplicated로 수정하는 것이 맞는 것 같네요.!

Hosick commented on src/main/java/com/flab/shoeauction/user/domain/User.java:
```diff
@@ -0,0 +1,44 @@
+import java.time.LocalDateTime;
+
+@Getter
+@Setter
```
도메인 클래스에 Setter 메소드를 작성하면 객체의 값이 언제 어디서 변해야 하는지 알수 없어지므로 파악하기 어려울 것이라고 생각되는데 대신 필드값 변경이 필요할 때 명확한 목적을 가진 메소드를 추가하는게 낫지 않을까요?

kihyuk-jeong replied:
네 `@Setter` 대신 `@Builder`와 목적에 따른 추가 메소드를 작성하는것 이 나을 것 같습니다. `@Setter`는 삭제하도록 하겠습니다!

Hosick commented on src/main/java/com/flab/shoeauction/user/domain/User.java:
```diff
@@ -0,0 +1,44 @@
+
+    private String phone;
+
+    private LocalDateTime createdAt;
```
앞으로 추가될 도메인들에도 필요한 필드같은데 JPA Auditing을 이용해서 따로 빼서 자동화하는게 좋다고 생각됩니다.
그리고 created 뿐만 아니라 modified도 있다면 좋을 것 같습니다.

kihyuk-jeong replied:
네. @Hosick 님의 의견처럼 BaseEntity를 하나 만들어서 공통적으로 관리해 주는 게 좋을 것 같네요. 수정일(modified)도 추가하는 게 좋을 것 같습니다.!

Hosick commented on src/main/java/com/flab/shoeauction/user/dto/UserDto.java:
```diff
@@ -0,0 +1,50 @@
+
+@Getter
+@Builder
+public class UserDto {
```
저는 단일책임원칙을 위배하지 않기 위해 DTO를 용도별로 조금 세분화했으면 좋겠다고 생각하는데 어떻게 생각하시나요?

kihyuk-jeong replied:
음. 저는 `UserDto`의 클래스명을 `SignUpDto`로 따로 분리할까 하다가도 너무 많은 DTO가 만들어질 것 같아서 DTO를 Inner class로 관리하는 방법을 알아보았는데 @Hosick님의 의견이 궁금하네요.
[DTO를 깔끔하게 관리하기](https://velog.io/@p4rksh/Spring-Boot에서-깔끔하게-DTO-관리하기) 이 글에대해서 어떻게 생각하시는지 의견이 궁금합니다.!

Hosick replied:
하나의 DTO 클래스에서 모두 처리하는 것보다 Inner class들을 만들어 사용하는 것이 더 깔끔하고 좋은 것 같네요!

Hosick commented on src/main/java/com/flab/soft/shoeauction/user/domain/User.java:
```diff
@@ -0,0 +1,45 @@
+    private LocalDateTime createdAt;
+
+    @Embedded
+    private Address address;
```
오븐으로 만든 뷰에는 없던 필드인데 따로 추가하신건가요? 같이 기능이나 뷰 설계를 조금 더 탄탄하게 통일하는데 시간을 조금 더 투자하는 것이 좋을 것 같습니다!

kihyuk-jeong replied:
네. 회원가입할때는 주소와 환급받을 계좌정보를 입력하지 않고 가입 후 마이페이지에서 추가하도록 설계하는 과정이기때문에 회원정보 Entity에 미리 작성해두었습니다. 회원가입과는 무관한 필드입니다.

Hosick commented:
leave a comment

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/controller/UserController.java:
```diff
@@ -0,0 +1,63 @@
+        return ResponseEntity.badRequest().build();
+    }
+
+    @PostMapping("/new")
```
https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

kihyuk-jeong replied:
`@PostMapping` 자체가 생성을 의미하기 때문에 `/new` url은 불필요할 것 같네요. 제거하는 게 맞는 것 같습니다.!

kihyuk-jeong committed:
refactor : 리뷰 반영 (1차)

1. Jpa Auditing을 사용하여 생성일과 수정일을 공통 관리하는 BaseEntity 추가
2. 네이밍 수정
3. 불필요한 Setter 메소드 제거

 회원가입 : #2

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/exception/CustomizeResponseEntityExceptionHandler.java:
```diff
@@ -0,0 +1,30 @@
+    @ExceptionHandler(PasswordMissMatchException.class)
+    public final ResponseEntity<Object> handlePasswordMissMatchException(Exception ex, WebRequest request) {
+        ExceptionResponse exceptionResponse =
+                new ExceptionResponse(LocalDateTime.now(), ex.getMessage(), request.getDescription(false));
```
이곳에서 로그도 남겨주시면 좋을 것 같네요~

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/EncryptionUtils.java:
```diff
@@ -0,0 +1,29 @@
+            byte[] passBytes = s.getBytes();
+            md.reset();
+            byte[] digested = md.digest(passBytes);
+            StringBuffer sb = new StringBuffer();
```
이거는 블로그에 있는 코드를 그대로 가져오신 것 같은데요~ `StringBuffer`와 `StringBuilder`의 차이는 무엇일까요?

kihyuk-jeong replied:
`Thread-safe`와 `not Thread-safe`의 차이로 알고있습니다. 메소드 내에서 지역변수로 사용되는 경우에는 thread-safe하기때문에 성능이 더 좋은 `StringBuilder`를 사용하는 것이 좋을 것 같습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/utils/SessionUtils.java:
```diff
@@ -0,0 +1,28 @@
+
+@Component
+@RequiredArgsConstructor
+public class SessionUtils {
```
이거는 세션을 이용한 인증을 처리하는 책임을 맡는 클래스인 것 같습니다. 이름도 그것에 맞게 지어줘야하지 않을까요?

kihyuk-jeong replied:
인증 세션 유틸리티라는 뜻의 `AuthenticationSessionUtils` 로 클래스명 수정하겠습니다.

f-lab-dev replied:
세션을 이용해 인증과 관련된 **비지니스 로직**을 처리하는 클래스인것 같은데 `Utils`라는 이름이 맞을까요?

kihyuk-jeong replied:
`AuthenticationService` 로 정정하겠습니다.

kihyuk-jeong committed:
refactor : 리뷰 반영 (1차)

1. Jpa Auditing을 사용하여 생성일과 수정일을 공통 관리하는 BaseEntity 추가
2. 네이밍 수정
3. 불필요한 Setter 메소드 제거

회원가입 : #2

kihyuk-jeong committed:
[#2] refactor : 테스트코드 추가 및 리뷰 반영 (2차)

<description>
1. SHA256 알고리즘 StringBuilder로 변경
2. CustomizeResponseEntityExceptionHandler에 로그 추가
3. exception 관련 메소드 파라미터 구체적인 형태로 수정
4. SessionUtils 메소드 이름 '인증'과 관련지어 수정
5. User 클래스 기본생성자 accessLevel PROTECTED로 변경
6. 회원가입 실패 관련 controller 테스트코드 추가

<issue-number>
 [#2] 회원가입 구현

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/exception/CustomizeResponseEntityExceptionHandler.java:
```diff
@@ -12,17 +13,21 @@
     @ExceptionHandler(UserDuplicateException.class)
-    public final ResponseEntity<Object> handleUserDuplicateException(Exception ex, WebRequest request) {
+    public final ResponseEntity<Object> handleUserDuplicateException(UserDuplicateException ex, WebRequest request) {
+        log.error(ex.getMessage());
```
`.error` 메소드에는 예외를 통째로 로깅할 수 있는 옵션이 있습니다. 메세지만 로깅하는게 아니라 예외를 아예 로깅하는게 추후 문제 발생시에 디버깅하기가 쉽지 않을까요?

f-lab-dev commented on src/test/java/com/flab/shoeauction/user/service/SignUpServiceTest.java:
```diff
@@ -0,0 +1,82 @@
+        UserDto userDto2 = UserDto.builder()
+                .email("test123@test.com")
+                .password("test1234")
+                .password("test123")
```
일반적으로 비밀번호 확인은 프론트엔드에서 사용자의 실수를 방지하고자 쓰입니다. 서버에서 굳이 두개를 받을 필요가 있을까요?

kihyuk-jeong replied:
해당 테스트코드에 password를 두번 적은것은 오타였는데 프론트쪽에서 비밀번호 검증을 해준다면 서버에서 두개를 받는것은 낭비일것같습니다!  `confirmPassword` 관련 dto와 검증하는 로직 수정/삭제 처리하겠습니다.

kihyuk-jeong committed:
[#2] refactor : 리뷰 반영(3차)

<description>
1. AuthenticationSessionUtils를 비즈니스 로직이라는 의미에 맞게 네이밍
(AuthenticationSessionUtils -> AuthenticationService)
2. 예외 log 추가
3. 회원가입시 비밀번호 확인 삭제
(대부분 프론트엔드에서 처리하므로 서버에서 비밀번호를 두번 받아서 검증할 필요 없음)

<issue-number>
 #2

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/exception/CustomizeResponseEntityExceptionHandler.java:
```diff
@@ -18,18 +17,9 @@
     @ExceptionHandler(UserDuplicateException.class)
     public final ResponseEntity<Object> handleUserDuplicateException(UserDuplicateException ex, WebRequest request) {
-        log.error(ex.getMessage());
+        log.error("UserDuplicateException", ex);
```
메소드는 적절하게 사용하셨는데 예외 메세지가 상세하지가 않네요~

kihyuk-jeong committed:
[#2] refactor : 리뷰반영

<description>
1. exception log 수정
2. AuthenticationService -> SmsCertificationService 이름 변경
3. 인증번호 세션 유효기간 3분으로 설정

<issue-number>
 #2

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/service/SignUpService.java:
```diff
@@ -53,7 +52,7 @@ public void saveAuthenticationNumber() {
-        authenticationService.setCertificationSession(stringBuffer.toString());
-        log.info(authenticationService.getCertificationSession());
+        smsCertificationService.setCertificationSession(stringBuffer.toString());
+        log.info(smsCertificationService.getCertificationSession());
```
이 로그는 info로 설정하신 이유가 무엇일까요?

kihyuk-jeong replied:
테스트단계에 문자메시지를 실제로 보내서 테스트하면 건당 30원의 이용료가 부과되어서  인증번호를 확인하기 위해 찍어둔 로그인데 상태변경과 같은 정보성 로그가 아니기때문에 info로 설정하는것은 잘못된것 같습니다. 실제 인증번호를 전송하는 API 추가하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/exception/CustomizeResponseEntityExceptionHandler.java:
```diff
@@ -17,7 +17,7 @@
     @ExceptionHandler(UserDuplicateException.class)
     public final ResponseEntity<Object> handleUserDuplicateException(UserDuplicateException ex, WebRequest request) {
-        log.error("UserDuplicateException", ex);
+        log.error(ex.getMessage(), ex);
```
`ex.getMessage()`는 `ex`가 출력되면서 같이 보일텐데 그럼 중복으로 출력하는게 아닐까요?
개발자는 고민이 중요합니다. 어떻게 해야 이 로그가 디버깅할때 더 용이하게 쓰이게 만들지 귀찮고 시간이 아깝게 느껴지더라도 계속 고민하셔야합니다

kihyuk-jeong replied:
로그에 exception이 발생한 url과 발생 시간을 추가하여 예외 발생 원인과 시간을 개발자가 쉽게 파악하도록 수정하겠습니다.추가적으로 예외 발생시 리턴되는 body값에는 예외메시지만 전송해도 될 것같아서 불필요한 `ExceptionResponse` 클래스는 삭제하겠습니다.

kihyuk-jeong committed:
<label>:<title>
[#2]  refactor : 리뷰반영 및 문자메시지 전송 추가

<description>
1. 자주사용하는 HttpStatus 상수 생성
2. 이메일 중복과 닉네임 중복 exception 분리하여 예외의 원인을 정확히 파악하도록 구현
3. exception이 발생한 시점과 발생한 url을 log로 출력하여 exception이 발생한 시간과 url, 원인을 알 수 있도록 구현
   (리턴되는 body에는 exception message만 전달하도록 수정)
4. 암호화 알고리즘 추상화 :  다른 암호화 알고리즘으로 변경될때 유지보수가 용이하도록 추상화
5. 실제 문자메시지 전송 구현 : SmsSendService
 (문자메시지 전송 관련 의존성 추가 : coolSMS)
6. REST API 규약에 맞도록 중복 체크 url 수정

<issue-number>
 #2

kihyuk-jeong committed:
[#2] refactor : 비즈니스 로직 분리 / sms 발송 api 추가

1. 회원가입 관련 비즈니스 로직과 SMS문자 인증 관련 비즈니스 로직 분리
2. coolSMS API를 추가하여 실제 sms인증과 동일한 서비스 구현

<issue-number>
 #2

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/service/SmsSendService.java:
```diff
@@ -0,0 +1,35 @@
+            System.out.println(obj.toString());
+        } catch (CoolsmsException e) {
+            System.out.println(e.getMessage());
+            System.out.println(e.getCode());
```
이 메소드는 절대 쓰면 안돼는 메소드입니다. 왜그럴까요?

kihyuk-jeong replied:
`Synchronized`가 적용되어 있기 때문에 성능 저하가 일어나기 때문입니다.
에러 메시지는 log로 처리하겠습니다.

f-lab-dev replied:
안좋은 이유가 하나 더 있었습니다~

kihyuk-jeong replied:
`System.out.println`을 사용할 경우 I/O 작업이 일어나기 때문에 커널 CPU를 불필요하게 많이 점유하게 됩니다. 또한 `Synchronized`로 인해 오버헤드가 발생하여 성능 저하가 일어나며, log level이 존재하지 않기 때문에 개발 단계에서만 사용되므로 의미 없는 디버그용 로그를 프린트하기 위해 서버싀 리소스와 디스크를 낭비하게 됩니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/service/SmsSendService.java:
```diff
@@ -0,0 +1,35 @@
+        params.put("to", phoneNumber);    // 수신전화번호
+        params.put("from", "01020180103");    // 발신전화번호
+        params.put("type", "SMS");
+        params.put("text", "shoe-action 휴대폰인증  : 인증번호는" + "["+certificationNumber+"]" + "입니다.");
```
checkstyle 혹시 적용하셨을까요?

kihyuk-jeong replied:
`google checkstyle`로 일관성있게 적용하겠습니다.!

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/service/SmsSendService.java:
```diff
@@ -0,0 +1,35 @@
+    public void sendMessage(String phoneNumber, String certificationNumber) {
+
+        String api_key = "";
+        String api_secret = "";
```
예제 코드를 그대로 복붙해오고 그 코드의 품질에 대해 확인 안해보는 습관은 정말 안좋은 습관입니다~

kihyuk-jeong replied:
이전 프로젝트에서 사용했던 코드였는데 당시에 성능을 고려하지 않고 구현했던 코드를 그대로 사용한 것 같습니다. 예제 코드도 다시 한번 살펴보는 습관 들이겠습니다 감사합니다.!

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/service/SignUpService.java:
```diff
@@ -1,58 +1,66 @@
+
+  public boolean emailDuplicateCheck(String email) {
+    return userRepository.existsByEmail(email);
+  }
```
여기는 indent를 공백 2개로 주셨는데 이렇게 정하신 이유가 있으실까요?

kihyuk-jeong replied:
`Google checkStyle` 기본 적용 값이 indent가 2라서 전체적으로 이렇게 적용된 것 같습니다.
@Hosick 님과 의논해서 수정하겠습니다.

kihyuk-jeong replied:
indent , tap size 4로 수정하였습니다. 다음 PR부터 적용하겠습니다.!

kihyuk-jeong committed:
[#2] refactor : 리뷰반영

<description>
1. sendMessage 메소드 수정 (exception 처리)
2. goole checkStyle 적용

<issue-number>
 #2

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/utils/UserConstants.java:
```diff
@@ -2,6 +2,8 @@
 
-  public static final int NUMBER_GENERATION_COUNT = 6;
+  public static final String API_KEY = "NCSBVON9BZGTHH2U";
+  public static final String API_SECRET = "CLFRQNF13AWV5VHXSMTFWZKIZRJISLK7";
```
이 값은 민감한 값인것같은데 여기에 하드코딩해도 괜찮을까요?

kihyuk-jeong replied:
충전된 금액이 없어서 넣었는데 개인 정보이므로 노출하지 않는 것이 좋아 보이네요.!

f-lab-dev replied:
이런 값은 커밋내역에도 남기때문에 공개된 저장소에서는 해커의 표적이 됩니다. 모두 재발급받으시는게 좋겠네요

kihyuk-jeong committed:
[#2] update : 휴대폰 인증 로직 변경

<description>
1. 기존 세션을 통해 인증하던 방법에서 concurrenthashmap 이용한 방법으로 변경

<issue-number>

kihyuk-jeong committed:
update : 휴대폰 인증 로직 변경

1.기존 session 방식에서 concurrenthashmap을 사용하는 방식으로 변경

kihyuk-jeong committed:
fixed untracked files

kihyuk-jeong commented:
제가 실수로 gitigonre 적용을 풀어버려서 불필요한 파일들이 올라갔다가 다시 커밋하느라 수정하지 않은 내용들이 올라갔습니다 ㅠㅠ  기존 SmsCertificationService 에서 인증번호를 세션으로 관리하는 방법에서 `ConcurrentHashMap`을 사용한 로직만 변경된 내용입니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/service/SmSCertificationService.java:
```diff
@@ -0,0 +1,63 @@
+@Slf4j
+public class SmSCertificationService {
+
+    private final ConcurrentHashMap<String, String> certificationInformation = new ConcurrentHashMap<>();
```
이렇게 서버에 정보를 저장하면 이 서버는 stateless하지 않은 서버가 됩니다. 그러면 어떤 단점이 생길까요?

kihyuk-jeong replied:
stateful 하면 서버가 클라이언트의 이전 내용을 모두 보존하고 있어야 하기 때문에 캐시의 효율적인 사용과 scale-out 방식을 이용한 서버 확정에 문제가 발생합니다. 또한 HTTP의 기본 통신 원칙이 stateless이므로 stateless 하지 않게 구현하는 것은 올바르지 못한 구현 방법인 것 같습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/service/SmSCertificationService.java:
```diff
@@ -0,0 +1,63 @@
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class SmSCertificationService {
```
추가적으로 SMS인증과 회원가입은 다른 이슈입니다. 다른 브랜치로 분리해서 개발하는게 낫지 않았을까요?

kihyuk-jeong replied:
회원가입을 과정에서 진행되는 인증이기 때문에 회원가입과 함께 분류했는데 SMS 인증 자체가 회원가입 로직보다 더 커지는 것 같네요. 따로 분리하는 게 맞는 것 같습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/user/utils/UserConstants.java:
```diff
@@ -2,8 +2,7 @@
-  public static final String API_SECRET = "CLFRQNF13AWV5VHXSMTFWZKIZRJISLK7";
-  public static final String CERTIFICATION_SESSION_KEY = "certificationNumber";
+  public static final String API_KEY = "";
+  public static final String API_SECRET = "";
```
이렇게하면 서버에 배포하고나서도 제대로된 값을 받지 못하겠네요~
이런 값은 외부에서 주입받도록 하는게 낫지 않을까요?

kihyuk-jeong replied:
외부 API key 또는 secret key는 하드 코딩보다는 application.yml에 설정해놓고 주입받는 편이 좋을 것 같습니다. sms 인증 구현할 때 반영하겠습니다.

f-lab-dev replied:
config 외부주입도 다른 이슈에 묻어가기보단 별도의 이슈를 만들어서 작업해주세요~ 이거는 일단 로그인 구현 각자 하는것들 끝내고서 생각하죠

kihyuk-jeong committed:
[#2] chore : sms 인증 브랜치 분리
